### PR TITLE
Simplify Ca and CaReconciler logic

### PR DESCRIPTION
**Pull request by katheris**
### Type of change

_Select the type of your PR_

- Refactoring

### Description

Simplify Ca and CaReconciler logic:
1. Remove code in CaReconciler that is never being executed.
2. Move Secret existence checking from CaReconciler to Ca.

Here are fuller explanations of why I did the refactors.

1
Today `logRenewalState()` in `Ca.java` is only called from inside [`shouldCreateOrRenew()`](https://github.com/strimzi/strimzi-kafka-operator/blob/main/operator-common/src/main/java/io/strimzi/operator/common/model/Ca.java#L688), and `shouldCreateOrRenew()` is only called from [`createRenewOrReplace()`](https://github.com/strimzi/strimzi-kafka-operator/blob/main/operator-common/src/main/java/io/strimzi/operator/common/model/Ca.java#L533) and only when `generateCa` is `true`. Therefore any code in `logRenewalState()` that is only executed when `generateCa` is false is never executed. To further simplify the code I removed `logRenewalState()` completely and moved the code that is executed inside `shouldCreateOrRenew()`.

2
Today `createRenewOrReplace()` in `Ca.java` is only called (excluding test classes) from [`CaReconciler.java`](https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/CaReconciler.java#L269), once for the ClusterCA and once for the ClientsCA. Prior to that call we call [`checkCustomCaSecret()`](https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/CaReconciler.java#L262) to verify the Secret exists. Then inside `Ca.java` we have additional logic to handle the case where the Secrets are null. I moved the verification logic into `Ca.java` to simplify the code.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

**Review by ppatierno — APPROVED** [review]
LGTM. Thanks!

**Review by tombentley — APPROVED** [review]

**Comment by ppatierno** [comment]
@scholzj do you have any plan to review this one or I can proceed by merging it? Thanks!

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/CaReconcilerTest.java [hunk 1]
```diff
@@ -327,7 +327,7 @@ public void testReconcileCasWhenCustomCertsAreMissingThrows(Vertx vertx, VertxTe
         reconcileCa(vertx, certificateAuthority, certificateAuthority)
             .onComplete(context.failing(e -> context.verify(() -> {
                 assertThat(e, instanceOf(InvalidResourceException.class));
-                assertThat(e.getMessage(), is("Cluster CA should not be generated, but the secrets were not found."));
+                assertThat(e.getMessage(), is("cluster-ca should not be generated, but the secrets were not found."));
```
Comment (scholzj) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/CaReconcilerTest.java:330]: In the docs we call it `Cluster CA` So why not keep it in the exception message?

File: operator-common/src/main/java/io/strimzi/operator/common/model/Ca.java [hunk 2]
```diff
@@ -525,12 +524,15 @@ public void createRenewOrReplace(
         int caKeyGeneration = keyGeneration();
 
         if (!generateCa) {
-            certData = caCertSecret != null ? caCertSecret.getData() : emptyMap();
-            keyData = caKeySecret != null ? singletonMap(CA_KEY, caKeySecret.getData().get(CA_KEY)) : emptyMap();
+            if (caCertSecret == null || caKeySecret == null)   {
+                throw new InvalidResourceException(this + " should not be generated, but the secrets were not found.");
+            }
```
Comment (scholzj) [inline operator-common/src/main/java/io/strimzi/operator/common/model/Ca.java:529]: An empty line after the block would be nice. But I think it is strange and error-prone to have a validation of something passed to a constructor and stored in the class of a class as part of a method that is called separately. I think it is fine not to have it outside, but it should be done in the constructor and not in some method called only later.

**Review by scholzj — COMMENTED** [review]
I left some nits.

File: operator-common/src/main/java/io/strimzi/operator/common/model/Ca.java [hunk 3]
```diff
@@ -685,50 +687,15 @@ && certNeedsRenewal(currentCert)) {
...
-            } else if (caCertSecret == null) {
-                LOGGER.warnCr(reconciliation, "The certificate (data.{}) in Secret {} and the private key (data.{}) in Secret {} in namespace {} " +
-                                "needs to be configured with a Base64 encoded PEM-format certificate. " +
-                                "Alternatively, configure Kafka.spec.tlsCertificates.generateCertificateAuthority=true in the Kafka resource with name {} in namespace {}.",
-                        CA_CRT.replace(".", "\\."), this.caCertSecretName,
-                        CA_KEY.replace(".", "\\."), this.caKeySecretName, namespace,
-                        clusterName, namespace);
-            }
+            case NOOP ->
+                    LOGGER.debugCr(reconciliation, "{}: The CA certificate in secret {} already exists and does not need renewing", this, caCertSecretName);
```
[the remark this answers is not in the document]
Comment (katheris) [reply operator-common/src/main/java/io/strimzi/operator/common/model/Ca.java:696]: From looking at the code I think the reason for the inconsistency is that we have given a more specific message that what we would have in NOOP. Here we are specifically logging the behaviour as it relates to certs needing renewal, however the NOOP enum also implies the key doesn't need replacing.

It looks like nothing else calls this preDescription at the moment, but I guess we risk printing out a confusing error message later if someone starts using it. Or we would have to have a generic message in the NOOP enum which might not be as helpful here.

I can remove preDescription and have each type with a separate case here if you'd prefer?

[the remark this answers is not in the document]
Comment (scholzj) [reply operator-common/src/main/java/io/strimzi/operator/common/model/Ca.java:696]: > guess we risk printing out a confusing error message later if someone starts using it.

Well, that is a problem for whoever uses it in the future - if anyone ever uses it somewhere else. So I would not be worried about that here.

> I can remove preDescription and have each type with a separate case here if you'd prefer?

That would be fine for me as well. If the method cannot be really used everywhere, than it IMHO does not make sense to have it in the Enum.

File: operator-common/src/main/java/io/strimzi/operator/common/model/Ca.java (hunk 2 above)
Comment (katheris) [reply operator-common/src/main/java/io/strimzi/operator/common/model/Ca.java]: I've moved it to the constructor

File: operator-common/src/main/java/io/strimzi/operator/common/model/Ca.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (katheris) [reply operator-common/src/main/java/io/strimzi/operator/common/model/Ca.java]: Looking again at the method I think it's fine to change the message to be more generic. We mention the key in the CREATE message, so I updated the NOOP preDescription and changed the switch to use that

**Review by scholzj — APPROVED** [review]

**Pull request merged**